traits/observers/_i_observer.py

```diff
@@ -29,3 +44,104 @@ def __hash__(self):
     def __eq__(self, other):
         """ Return true if this observer is equal to the given one."""
         raise NotImplementedError("__eq__ must be implemented.")
+
+    @property
+    def notify(self):
+        """ A boolean for whether this observer will notify for changes.
+        """
+        raise NotImplementedError("notify property must be implemented.")
+
+    def iter_observables(self, object):
```

Comment:
I find the `iter_*` interface really confusing. I had to look at interface implementations to understand what exactly each one is iterating over just to find that `iter_observables` and `iter_extra_graphs` almost always return only one object. And with just having this interface description it's really hard to understand what `add_or_remove_notifiers` is doing as well.

I don't know how to fix this or even if it's worth fixing. Just wanted to bring this up, maybe someone else has any ideas.

Reply:
Thank you for bringing this up. I am also not sure how to fix this, maybe it just needs a different docstring?

Let me try to explain: For these `iter_*`, the interface is asking the implementation to "give me zero or many observables", therefore these are generators.

For example, if we want to observe traits that have a specific metadata, this `iter_observables` may yield nothing, if no traits have the required metadata, or many observables, if an instance of `HasTraits` has more than one trait that has the required metadata. Similarly for `iter_objects`.

For `iter_extra_graphs`, so far it is only needed for adding an observer for the `trait_added` event.  Some observers don't need the handling `trait_added` event, and their `iter_extra_graphs` will yield nothing. Other observer will only yield one extra graph for handing `trait_added`.
An alternative to this is to define a method that will return an instance of ObserverGraph or None for no graphs. But then `add_or_remove_notifiers` does not need to care whether there must be at most one extra graphs. It can deal with many graphs just fine. So a generator to ask for zero or many things seem to be the most flexible (then `add_or_remove_notifiers` also doesn't need to check for `None`).

Reply:
> "give me zero or many observables things"

To be precise, I mean a minimum of zero things, with no maximum bound. The count of one is allowed here too.

Reply:
Maybe some examples would help:

`iter_obserbables` yielding one thing for observing a specifically named trait:
https://github.com/enthought/traits/blob/1ea851ad118d054096ee13cd1e57a6d592fe81a2/traits/observers/_named_trait_observer.py#L80-L103

`iter_observables` yielding potentially nothing, but maybe many things when observing traits filtered by some callable (e.g. existence of a metadata):
https://github.com/enthought/traits/blob/1ea851ad118d054096ee13cd1e57a6d592fe81a2/traits/observers/_filtered_trait_observer.py#L64-L91

`iter_extra_graphs` yielding nothing for observing items on a list, because there is no "add_trait" action on a TraitList(TraitListObject to be precise):
https://github.com/enthought/traits/blob/1ea851ad118d054096ee13cd1e57a6d592fe81a2/traits/observers/_list_item_observer.py#L202-L215

`iter_extra_graphs` yielding one thing for handling `HasTraits.add_trait` as the observer is given an instance of `HasTraits`:
https://github.com/enthought/traits/blob/1ea851ad118d054096ee13cd1e57a6d592fe81a2/traits/observers/_named_trait_observer.py#L171-L190

Reply:
Examples would definitely help - looking at implementation is how I eventually understood what each generator is for. I'm just wondering if they might become a maintenance problem in the future 🤔

Reply:
I have expanded the docstring to mention some examples. See what you think. I hope to keep it generic enough so that examples don't get confused as rules.

I think it would also help when the concrete implementations are added to the code base, I can't do that yet because they are going to depend on the notifier that is still being reviewed.

Reply:
New docstrings look good! Thanks for that.

As per out discussion on slack, I was trying to think of a better phrasing for the first sentence of `iter_objects` docstring but came up with nothing. I think warnings on implementations that deal with CTrait should be enough.

Reply:
SGTM!  Thank you for hashing this out!